draft-rescorla-tls-esni.md

```diff
@@ -606,8 +606,57 @@ lie. In this design, the attacker would need to be able to find a
 Z which would expand into a key that would validly AEAD-encrypt
 a message of his choice, which should be intractable (Hand-waving alert!).
 
+# Alternative SNI Protection Designs
 
-# Alternate Encryption Design
+Alternative approaches to encrypted SNI may be implemented at the TLS or 
+application layer. In this section we describe several alternatives and discuss 
+drawbacks in comparison to the design in this document.
+
+## TLS-layer
+
+### TLS in Early Data
+
+In this variant, TLS Client Hellos are tunneled within early data payloads
+belonging to outer TLS connections established with the client-facing server. This 
+requires clients to have established a previous session -— and obtained PSKs —- with 
+the server. The client-facing server decrypts early data payloads to uncover Client Hellos
+destined for the hidden server, and forwards them onwards as necessary. All Server
+Hello messages generated in response by hidden servers are forwarded directly to
+client -- unmodified. This avoids double encryption of TLS handshake messages.
```

Comment:
Thank you for explicitly stating that only Client Hello is sent as 0-RTT data.

Maybe we might want to change the sentence starting from "All Server Hello ..." to something like "Rest of the exchanges between the client and the hidden server, starting from Server Hello, are forwarded by the client-facing server -- unmodified.", to clarify that exchanges following Server Hello are also forwarded directly.

Other than that, the PR looks ready for merge to me.

Reply:
The reason I suggested the change is because the paragraph seemed to be unclear to me on whether or not the handshake messages and application data that follows Server Hello is double-encrypted, because it only discusses about Client Hello and Server Hello explicitly.

Reply:
@kazuho Should be more clear now!